pkg/kv/kvnemesis/generator.go

```diff
@@ -500,6 +514,100 @@ func randPutExisting(g *generator, rng *rand.Rand) Operation {
 	return put(key, seq)
 }
 
+func randAddSSTable(g *generator, rng *rand.Rand) Operation {
+	ctx := context.Background()
+
+	sstTimestamp := hlc.MinTimestamp // replaced via SSTTimestampToRequestTimestamp
+	numKeys := rng.Intn(16) + 1      // number of point keys
+	probReplace := 0.2               // probability to replace existing key, if possible
+	probTombstone := 0.2             // probability to write a tombstone
+	asWrites := rng.Float64() < 0.2  // IngestAsWrites
+
+	// AddSSTable requests cannot span multiple ranges, so we try to fit them
+	// within an existing range. This may race with a concurrent split, in which
+	// case the AddSSTable will fail, but that's ok -- most should still succeed.
+	ranges := make([]string, 1, len(g.currentSplits)+1) // first range starts at ""
+	for key := range g.currentSplits {
+		ranges = append(ranges, key)
+	}
+	sort.Strings(ranges)
+
+	var rangeStart, rangeEnd string
+	rangeIdx := rng.Intn(len(ranges))
+	rangeStart = ranges[rangeIdx]
+	if rangeIdx+1 < len(ranges) {
+		rangeEnd = ranges[rangeIdx+1]
+	}
+	inRange := func(key string) bool {
+		return key >= rangeStart && (key < rangeEnd || rangeEnd == "")
+	}
+
+	// Generate keys first, to write them in order and without duplicates.
+	//
+	// TODO(erikgrinaker): For now, only ingest point keys. We currently don't
+	// ingest range keys in production code, although we soon will.
+	curKeys := make([]string, 0, len(g.keys)/len(ranges))
+	for key := range g.keys {
+		if inRange(key) {
+			curKeys = append(curKeys, key)
+		}
+	}
+
+	sstKeys := []string{}
+	sstKeysMap := map[string]struct{}{}
+	for len(sstKeys) < numKeys {
+		var key string
+		if len(curKeys) > 0 && rng.Float64() < probReplace {
+			key = curKeys[rng.Intn(len(curKeys))]
+		} else {
+			for key == "" || !inRange(key) {
+				key = randKey(rng)
+			}
+		}
+		if _, ok := sstKeysMap[key]; !ok {
+			sstKeysMap[key] = struct{}{}
+			sstKeys = append(sstKeys, key)
+		}
+	}
+	sort.Strings(sstKeys)
+
+	sstSpan := roachpb.Span{
+		Key:    roachpb.Key(sstKeys[0]),
+		EndKey: roachpb.Key(sstKeys[len(sstKeys)-1]).Next(),
+	}
+
+	seq := g.nextSeq()
+	sstValueHeader := enginepb.MVCCValueHeader{}
+	sstValueHeader.KVNemesisSeq.Set(seq)
```

Comment:
It seems like it would be cleaner to construct the SST in the applier instead (when you send out the batch). This avoids fragmenting the sequence numbers: right now you store them into the SST here, and then in the applier, still set the seqno onto the RequestHeader, which isn't strictly necessary (probably just to avoid tripping some assertions?)

Moving SST creation into applier gives you one place in which to explain how the plumbing works for this request type.

Reply:
I see that you're iterating through the SST when formatting, which makes sense - would be awkward to have to write more formatting code than is truly necessary. I'm fine with this as-is, please add comments in both places that deal with seqnos to guide future confused readers.

Reply:
We generate the seqno in the generator for the other operations too, no, and then pass them up to the applier via `Operation.Seq`? I don't really see how this is different from what e.g. `Put` does, except for how they're plumbed through to storage.

You're right that we don't really need to pass the seqno in the RequestHeader, I just did it for consistency with other operations. I'm fine to drop it too. Will add comments.